### PR TITLE
HDDS-10632. Handle inconsistent read issue for hsync keys after lease recovery.

### PR DESCRIPTION
**Pull request by ashishkumar50**
## What changes were proposed in this pull request?

In this PR, In some cases we are considering last two block to get the actual length from DN and update OM fileTable during lease recovery.
This is done as there may be some corner case when hsync is not called on last block and lease recover is called and also penultimate block is not completely filled because of various reason like CLOSE container. In this case penultimate block length is not correct in openFileTable as well as fileTable.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10632

## How was this patch tested?

New integration test

### CODE REVIEW

File: hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java [hunk 1]
```diff
@@ -156,21 +157,55 @@ public boolean recoverLease(Path f) throws IOException {
...
+      if (openKeyLocationSize > 1) {
+        openKeyFinalBlock = openKeyLocationInfoList.get(openKeyLocationSize - 1);
+        openKeyPenultimateBlock = openKeyLocationInfoList.get(openKeyLocationSize - 2);
+      } else if (openKeyLocationSize > 0) {
+        openKeyFinalBlock = openKeyLocationInfoList.get(0);
+      }
+      // Finalize the final block and get block length
       try {
-        block.setLength(getAdapter().finalizeBlock(block));
+        // CASE 1: When openFileTable has more block than fileTable
```
Comment (jojochuang) [inline hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java:183]: does the test case cover all three cases?

**Review by jojochuang — COMMENTED** [review]
Hi Ashish thanks for the PR. I'm half way through just want to leave a quick comment. I'll finish it up asap.

File: hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java (hunk 1 above)
Comment (jojochuang) [reply hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java:183]: the rest looks good.

**Review by jojochuang — COMMENTED** [review]
also the code looks so complex now it would be great to refactor out the common code into another helper method.

File: hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java (hunk 1 above)
Comment (ashishkumar50) [reply hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java]: Thanks for the review, Case1 and 2 is covered in the new and existing test case. One more test case I have added now to cover Case3.

**Comment by ashishkumar50** [comment]
> also the code looks so complex now it would be great to refactor out the common code into another helper method.

Yes code was duplicated multiple places, Refactored now to keep in only one place.

**Review by jojochuang — APPROVED** [review]
LGTM

**Pull request merged**